### ISSUE TITLE
BuildPackages.sh vs. InstPackages.sh

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#737]
So one the one hand, INSTALL.md points people to http://www.gap-system.org/Download/InstPackages.sh and http://www.gap-system.org/Download/InstPackages32.sh

OTOH, we recently started bundling an arguably better (as it is more robust) script `bin/BuildPackages.sh` with GAP -- which, however, is nowhere advertised.

We should clean this up. I suggest the following:
1. Resolve issue #736
2. Update INSTALL.md to reference it, instead of the `InstPackages.sh` / `InstPackages32.sh` scripts.
3. On the server, replace `InstPackages.sh` and `InstPackages32.sh`  either by a copy of `bin/BuildPackages.sh`, or (probably better) by a simple script which just prints something like this: "This script is obsolete, please use the bin/BuildPackages.sh script bundled with GAP instead."

**Comment by alex-konovalov** [comment]
I agree to switch to BuildPackages.sh, but in its current form it tries to build each package having Makefile even if the package has no binaries and used make e.g. to build manuals and/or wrap archives. I think it should explicitly blacklist packages not compatible with it. See also https://github.com/gap-system/gap-distribution/issues/17

**Comment by fingolfin** [comment]
I will work on this now.

**Comment by fingolfin** [comment]
@alex-konovalov Could you please give an example of a package that should be blacklisted? (Even better: a full list ;-)

**Comment by ChrisJefferson** [comment]
@alex-knovalov : Yes, this would be nice. My only personal preference is I would strongly prefer we don't have an opt-in list -- because then we have to remember to add every new package as it comes along. Hopefully over time this script will get smaller and smaller, as packages move to a better standard.

**Comment by alex-konovalov** [comment]
@ChrisJefferson : yes, agree - the whole point of this is to get rid of opt-in list. @fingolfin I will provide a full list of packages for opt-out list now!

**Comment by alex-konovalov** [comment]
It happens that 46 packages from GAP 4.8.3 distribution have Makefiles. To the contrary, https://github.com/gap-system/gap-distribution/blob/master/testing/stable/InstPackages.sh builds 25 packages, and also calls `chmod 1777 datagens dataword`  for AtlasRep package. Hence, the list of packages to opt-out is:
```
4ti2Interface
AutoDoc-2016.03.08
CAP-2016.02.19
Convex
ExamplesForHomalg
GaussForHomalg
GeneralizedMorphismsForCAP
GradedModules
GradedRingForHomalg
HomalgToCAS
IO_ForHomalg
LinearAlgebraForCAP
LocalizeRingForHomalg
MatricesForHomalg
ModulePresentationsForCAP
Modules
RingsForHomalg
SCO/makefile
ToolsForHomalg
ToricVarieties
gbnp
genss
homalg
qaos-1.2
recog-1.2.5
recogbase-1.2.5
```
I am not sure what to do with Simpcomp:
```
 * (Optional) Run *"chmod +x configure; ./configure; make; make install; make clean"* (This step is only required if you want to use 
   *SCReduceComplexFast()*). You may need admin rights to perform this step.
```

**Comment by fingolfin** [comment]
Errr, but guys, `BuildPackages.sh` does not build any of these packages. Sure, it says `==== Building AutoDoc-2016.03.08/`, but it doesn't actually *do* anything: If you look at the script, it only does something if there is either a `configure` script present, or else in a special case (for `atlasrep`).

So perhaps what you are asking for is to suppress the message `==== Building PACKAGE` for those packages? That could be done, but it would require some effort to get the correct and uniform output then, considering the special cases. So I'd rather not do it. But we could change the message, e.g. to `==== checking package FOO`... ?

**Comment by ChrisJefferson** [comment]
I agree, how about something like (just made a little patch)

```
==== Checking Modules
No building required for Modules
==== Checking NormalizInterface-0.9.5
checking build system type... x86_64-apple-darwin15.4.0
```

**Comment by alex-konovalov** [comment]
@fingolfin aha, so that's actually better than in my initial impression. If we can get @ChrisJefferson's patch merged, I will be happy to update Jenkins to use BuildPackages.sh etc.

**Comment by fingolfin** [comment]
I don't know where @ChrisJefferson's patch is, but I would prefer if that was not committed directly, to avoid conflicts with my changes to that script, which will appear in a pull request shortly (you can in fact see a preview on #728).

@ChrisJefferson could you email me your diffs (possibly made using `git format-patch`, so I can retain attribution)? Or tell me in which repos and branch to find them?

**Comment by fingolfin** [comment]
Now that #740 is merged, the next step is something like https://github.com/gap-system/GapWWW/pull/19

The final step is to replace or remove http://www.gap-system.org/Download/InstPackages.sh -- but we should not be hasty about this, as the latest released version still refers to it. My suggestion would be to replace both InstPackages with the new BuildPackages, once we had a chance to test the latter some more.

**Comment by fingolfin** [comment]
Oh and as discussed with @alex-konovalov on IRC: the BuildPackage.sh script perhaps needs a new special case for anupq, which ensures that anupq always gets built in 32bit mode. We'll have to see about that.

**Comment by alex-konovalov** [comment]
Since this thread is still open, I will likely post further details for certain packages here.

First, ANUPQ standard tests are very unstable. For example, this is 32-bit build of GAP without GMP support on a 64-bit machine:
```
Testing anupq 3.1.4, test=/mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/tst/anupqeg.tst, all packages=false
*** glibc detected *** /mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq: free(): invalid pointer: 0x08ec985a ***
======= Backtrace: =========
/lib/libc.so.6[0xab0a15]
/lib/libc.so.6(cfree+0x59)[0xab4a89]
/mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq[0x804aa51]
/mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq[0x805a007]
/mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq[0x806b91d]
/lib/libc.so.6(__libc_start_main+0xdc)[0xa5cedc]
/mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq[0x8049101]
======= Memory map: ========
00a28000-00a43000 r-xp 00000000 08:11 2299403                            /lib/ld-2.5.so
00a43000-00a44000 r--p 0001a000 08:11 2299403                            /lib/ld-2.5.so
00a44000-00a45000 rw-p 0001b000 08:11 2299403                            /lib/ld-2.5.so
00a47000-00b9d000 r-xp 00000000 08:11 2299479                            /lib/libc-2.5.so
00b9d000-00b9f000 r--p 00156000 08:11 2299479                            /lib/libc-2.5.so
00b9f000-00ba0000 rw-p 00158000 08:11 2299479                            /lib/libc-2.5.so
00ba0000-00ba3000 rw-p 00ba0000 00:00 0 
00ba5000-00bcc000 r-xp 00000000 08:11 2299336                            /lib/libm-2.5.so
00bcc000-00bcd000 r--p 00026000 08:11 2299336                            /lib/libm-2.5.so
00bcd000-00bce000 rw-p 00027000 08:11 2299336                            /lib/libm-2.5.so
00bd0000-00bdb000 r-xp 00000000 08:11 2299333                            /lib/libgcc_s-4.1.2-20080825.so.1
00bdb000-00bdc000 rw-p 0000a000 08:11 2299333                            /lib/libgcc_s-4.1.2-20080825.so.1
08048000-08071000 r-xp 00000000 08:21 29197824                           /mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq
08071000-08072000 rw-p 00028000 08:21 29197824                           /mnt/disk2/hudson-slave/workspace/GAP-minor-release-test/GAPCOPTS/32build/GAPGMP/nogmp/GAPTARGET/packages/label/64bit/gap4r8/pkg/anupq-3.1.4/bin/x86_64-pc-linux-gnu-gcc-default32/pq
08ec9000-08eea000 rw-p 08ec9000 00:00 0 
f5939000-f7f60000 rw-p f5939000 00:00 0 
f7f60000-f7f93000 r-xp 00000000 08:11 1360154                            /usr/lib/sse2/libgmp.so.3.3.3
f7f93000-f7f94000 rw-p 00032000 08:11 1360154                            /usr/lib/sse2/libgmp.so.3.3.3
f7f94000-f7f95000 rw-p f7f94000 00:00 0 
f7fbf000-f7fc0000 rw-p f7fbf000 00:00 0 
ff849000-ff8e8000 rw-p 7ffffff5f000 00:00 0                              [stack]
ffffe000-fffff000 r-xp ffffe000 00:00 0 
```

**Comment by fingolfin** [comment]
@alex-konovalov Well, I don't think this is the right place for this, unless this is a regression compared to `InstPackages.sh`. But I suspect iti is not, i.e. the same problems already occured with `InstPackages.sh`.

In that case, the anupq problems should be reported and discussed on the anupq issue tracker alone; and this issue here can probably be closed.

**Comment by alex-konovalov** [comment]
@fingolfin You're most likely right - unless the fix is in the adjustments of the build instructions for anupq which were just inherited from InstPackages.sh.

There are also adjustments for Carat suggested by Franz in the Support mailing list - probably I should open new issue for that.

**Comment by ChrisJefferson** [comment]
There was a problem in how I copied anupq, I didn't realise it just doesn't work when built in 64-bit mode. Perhaps that should be fixed, but for now #800 forces it to always build in 32-bit mode.

**Comment by alex-konovalov** [comment]
@ChrisJefferson I am not sure that #800 helped, so let it waits in master. I will not include it until GAP 4.8.4 until further investigation.
